Add Macaulay2 lexer

d-torrance opened the pull request:
[Macaulay2](http://macaulay2.com) is a computer algebra system that is widely used for research in algebraic geometry and commutative algebra.  It comes with its own programming language, and it would be great if support were added to Pygments, e.g., for syntax highlighting Macaulay2 code in journal articles using [minted](https://ctan.org/pkg/minted?lang=en).

The lists of symbols are generated automatically using the script `make-M2-symbols.m2` in the Macaulay2 source.  See https://github.com/Macaulay2/M2/pull/2040.

d-torrance commented:
Macaulay2 1.18 was just released, so I've pushed a new commit with some updated symbols.

d-torrance commented:
I notice there's an "updated needed" label.  What update is requested?

Anteru commented:
> 
> 
> I notice there's an "updated needed" label. What update is requested?

Looks like I forgot to request changes when I reviewed. My bad -- sorry. Should be online now.

d-torrance commented on pygments/lexers/macaulay2.py:
```diff
@@ -0,0 +1,1661 @@
+        'root': [
+            (r'--.*$', Comment.Single),
+            (r'-\*[\w\W]*?\*-', Comment.Multiline),
+            (r'".*?"', String),
```
No idea!   Fixed in https://github.com/pygments/pygments/pull/1791/commits/6f8a0d1411a269e70d1b6fcdfe9269a0ff849eb5

d-torrance replied:
Actually, now I remember.  We want lazy matching here.  For example, each line below should be contain two separate comments/strings:

```m2
-* foo *- 2 -* bar *-
"foo" | "bar"
///foo/// | ///bar///
```

Removing my earlier fix.

Anteru replied:
Can you please make sure this is covered in the test case?

Anteru replied:
Also, this is a super error prone pattern I've seen a bunch of times. I'd suggest to implement a proper string state, because as soon as there is string escaping (which is usually inevitable) this becomes a mess. Basically you enter a string case when you see a ", you exit when you see it again, and within the string case you accept anything but ".

d-torrance replied:
Ok, I'll try that out!  Thanks for the suggestion.

d-torrance replied:
I've added states for strings and block comments, replacing the lazy matching, and have added accompanying tests, too.

d-torrance commented on pygments/lexers/macaulay2.py:
```diff
@@ -0,0 +1,1661 @@
+            (words(M2DATATYPES, prefix=r'\b', suffix=r'\b'), Name.Builtin),
+            (words(M2FUNCTIONS, prefix=r'\b', suffix=r'\b'), Name.Function),
+            (words(M2CONSTANTS, prefix=r'\b', suffix=r'\b'), Name.Constant),
+            (r'.', Text)
```
Done!  Fixed in https://github.com/pygments/pygments/pull/1791/commits/73559471639b0f03c5ee361ce29e27c0fcfc4876

d-torrance commented:
> > I notice there's an "updated needed" label. What update is requested?
> 
> Looks like I forgot to request changes when I reviewed. My bad -- sorry. Should be online now.

No problem!  I believe I've addressed the comments.

d-torrance commented:
Just pushed a couple new commits addressing the comments.

Anteru commented on pygments/lexers/macaulay2.py:
```diff
@@ -0,0 +1,1676 @@
+        'quote string' : [
+            (r'"', String, '#pop'),
+            (r'\\"', String),
+            (r'[\w\W]', String)
```
This should be `+` because right now you generate one token per character which is incredibly wasteful. Same in the `slash string`.

Anteru commented on pygments/lexers/macaulay2.py:
```diff
@@ -0,0 +1,1676 @@
+        ],
+        'block comment' : [
+            (r'\*-', Comment.Multiline, '#pop'),
+            (r'[\w\W]', Comment.Multiline)
```
Same situation as in the `quote string` below, please match using `+`, otherwise we get one token per character.

d-torrance commented:
I've pushed a new commit addressing the latest comments (plus another updating the test output).

Anteru commented:
Thanks, that looks much better now. I'll merge this soon, just waiting for any post 2.10 fallout which might require a point release.

d-torrance commented:
I've just pushed a new commit updating the symbols for the latest release of Macaulay2.

Anteru commented:
Can you please move the comment somewhere further down the file? The automated check expects each file to start with a docstring.

d-torrance commented:
> Can you please move the comment somewhere further down the file? The automated check expects each file to start with a docstring.

Done!

d-torrance commented:
Am I correct that the failed test is unrelated to this pull request?

```python
  =================================== FAILURES ===================================
  ____________________ test_long_unclosed_javascript_fragment ____________________
  
  lexer_html = <pygments.lexers.HtmlLexer>
  
      def test_long_unclosed_javascript_fragment(lexer_html):
          """unclosed, long Javascript fragments should parse quickly"""
      
          reps = 2000
          fragment = "<script type=\"text/javascript\">"+"alert(\"hi\");"*reps
          start_time = time.time()
          tokens = list(lexer_html.get_tokens(fragment))
  >       assert time.time() - start_time < MAX_HL_TIME, \
              'The HTML lexer might have an expensive error script case'
  E       AssertionError: The HTML lexer might have an expensive error script case
  E       assert (1638127016.9544926 - 1638127006.3483639) < 10
  E        +  where 1638127016.9544926 = <built-in function time>()
  E        +    where <built-in function time> = time.time
  
  /home/runner/work/pygments/pygments/tests/test_html_lexer.py:53: AssertionError
  =========================== short test summary info ============================
  FAILED tests/test_html_lexer.py::test_long_unclosed_javascript_fragment - Ass...
  ============ 1 failed, 3809 passed, 12 skipped in 390.88s (0:06:30) ============
  make: *** [Makefile:54: test] Error 1
```

Anteru commented:
This is PyPy3 failing and it's been very flaky recently, so yes, unrelated. @birkenfeld any concerns disabling PyPy3 until we find a way to make it not fail with 50% chance?

birkenfeld commented:
Sure!

d-torrance commented:
I noticed that this was bumped to the next release milestone.  I believe all the requested changes have been addressed -- am I missing something?

Anteru commented:
No, I just had to make the cut somewhere for the release (and that still went not as great as hoped for ...) It's firmly going into the next one, assuming the fallout from the current one is resolved (and to avoid creating a release branch for point releases I'm not merging this yet.)

d-torrance commented:
I've updated this branch so it no longer conflicts with `master` after the recent `languages.rst` updates and also added a `url` field.

Anteru commented:
Merged, thanks. I'll add a link back to the generation script so this can be re-created if necessary.

d-torrance commented:
> Dang it, this fails due to a missing copyright string. Can you please check once more and I'll merge this tomorrow? Sorry for leaving you hanging here more than half a year :(

The test passes after I bump the final copyright year from 2021 to 2022.  I just pushed a commit with the fix.
